### DISCUSSION TITLE
Provide filtering options to `gooseberry make` ?

### DISCUSSION THREAD

**Discussion by ngirard** [out-of-cheese-error/gooseberry#89] [Q&A] [answered]
I've started this discussion rather than an issue, because I should perhaps adapt my workflow to Gooseberry rather than the opposite here.

The filtering options to `search` and `view` and the proposed workflow make all perfect sense to me, until I try and bring my new annotations to my Obsidian vault.

I feel the need to digest my annotations from within Obsidian iteratively, i.e. dealing with them by chunks. For this, I would associate a tag to each research I'm performing, so that when working on a topic T, the related annotations are all tagged `T` -- or whatever equivalent.

So at that stage, I'd expect to be able to generate markdown files only for today's annotations, or for the topical tags related to my current work (like `T` in this discussion), instead of regenerating the whole stuff.

But perhaps I'm approaching it the wrong way ?

**Comment by ngirard** [comment]
(Loosely) related: #86.

**Comment by Ninjani** [comment]
This is actually something I've been thinking of implementing - I have the same issue, and I've been trying to get around it by using different hypothesis groups for different topics, but it's annoying to have to change the group in the sidebar constantly. The filters themselves are quite easy to implement - my main blocker is the index file. Would this file only contain the links generated by the current `make` run (easy), or will it need to have links based on all annotations (not hard to implement but would mean querying for all annotations on each run anyway)?

**Reply by ngirard** [reply]
> it's annoying to have to change the group in the sidebar constantly

Not to mention the risk of confusing groups when adding an annotation.
I'm doing it all the time, and once an annotation is added to a group, there are no way to move it to another one with the Firefox unofficial extension I'm using

**Reply by ngirard** [reply]
> My main blocker is the index file.
> Would this file only contain the links generated by the current `make` run (easy), or will it need to have links based on all annotations (not hard to implement but would mean querying for all annotations on each run anyway)?

I'm not sure I'm thinking clear enough this morning to delve into the question... so how about implementing this in two steps ?
1. When `make` is called with filters, no index is produced (and the whole kb dir is recreated)
2. Figure out something clever with the index later

**Reply by ngirard** [reply]
Another alternative: add another subcommand. `make_all` (or whatever) would be equivalent to the current `make`, and `make` (or whatever) would not produce a "knowledge base" but generate markdown files for the chosen subset of annotations. 
Perhaps 2 different use cases => 2 subcommands would be better ?

**Reply by Ninjani** [reply]
> I'm doing it all the time, and once an annotation is added to a group, there are no way to move it to another one with the Firefox unofficial extension I'm using

Have you tried `gooseberry move`? 😋  

>     1. When `make` is called with filters, no index is produced (and the whole kb dir is recreated)

Right, so this would work with Obsidian but not with mdbook (which needs a SUMMARY file) and I'd like for users to have a choice on which system they use.

**Reply by Ninjani** [reply]
> Another alternative: add another subcommand

This is interesting. An alternative would be to have `make` which generates files using optionally filtered annotations, and `index` which generates the index file (also using optionally filtered annotations if that's what's required). That way, for systems that don't actually need an index the user doesn't have to create one.

**Reply by ngirard** [reply]
> Right, so this would work with Obsidian but not with mdbook (which needs a SUMMARY file)

Aha. In this scenario, would it be conceivable that the SUMMARY file only contains links to the other files from the kb directory ?

**Reply by ngirard** [reply]
> An alternative would be to have `make` which generates files using optionally filtered annotations, and `index` which generates the index file (also using optionally filtered annotations if that's what's required). That way, for systems that don't actually need an index the user doesn't have to create one.

Another one (brainstorming time, yay): provide such `make --mdbook` / `make --obsidian` options so that Gooseberry is aware of what to do and saves the user from (at least some) plumbing decisions ?

**Reply by Ninjani** [reply]
> > An alternative would be to have `make` which generates files using optionally filtered annotations, and `index` which generates the index file (also using optionally filtered annotations if that's what's required). That way, for systems that don't actually need an index the user doesn't have to create one.
> 
> Another one (brainstorming time, yay): provide such `make --mdbook` / `make --obsidian` options so that Gooseberry is aware of what to do and saves the user from (at least some) plumbing decisions ?

Hmm, I'm not sure if I'm a fan of this as it would then need new flags for every new system and there are too many differences between systems (e.g different link types in obsidian, also org-mode is a different beast entirely!). I would prefer instead if there are discussion topics/wiki pages of user configurations for different systems like in #73 (since there are multiple possible setups anyway).

**Reply by ngirard** [reply]
> Have you tried `gooseberry move`?

Very useful feature indeed, thanks for pointing it out !

**Reply by ngirard** [reply]
> An alternative would be to have `make` which generates files using optionally filtered annotations, and `index` which generates the index file (also using optionally filtered annotations if that's what's required).

Would there be a way for `index` to reuse the filter options from the previous call to `make`, or should the same options be used for calling both `make` and `index` ?

**Reply by Ninjani** [reply]
> Would there be a way for `index` to reuse the filter options from the previous call to `make`, or should the same options be used for calling both `make` and `index` ?

I would say the latter simply because it would then be up to the user if they want to have links to nonexistent files or only the filtered annotations. Non-existent links could still be useful in some setups, to get an overall picture of the tag structure or URIs annotated for instance. What do you think?

**Reply by ngirard** [reply]
I'd say go for it !
I'll always be possible to do something more clever after experimenting with this first implementation.

**Reply by Ninjani** [reply]
Okay, on it! (Issue #90)

**Reply by ngirard** [reply]
Great, thanks !

**Reply by ngirard** [reply]
@Ninjani, wow, that was fast ! Nice job !
You seem to have become quite fluent in Rust ! Are you leveraging it in your bioinformatics-related activities ?

I've compiled a local binary with your recent work, and I wish to give it a try now.
(Getting your work was as easy as typing `gh pr checkout 91` from within my main local branch. Github CLI is really a nice tool, I should use it more often.)

So, I would like to test it in "real life" situation, for my daily tech monitoring.

I have a side question, before doing so: what would be the best way of simply "adding" a "whole" web page to Hypothes.is, without annotating it ? To rephrase, what would be the best Hypothes.is equivalent to "bookmarking a page" ? There are many cases when I'm bookmarking a page only to "process" it later, so if I could do the same using Hypothes.is without adding bookmarks, this might be an improvement to my workflow...

**Reply by Ninjani** [reply]
> Are you leveraging it in your bioinformatics-related activities ?

I wish! Hopefully I'll be in a position to do this soon, especially now that the python-rust interop is so smooth.

> what would be the best way of simply "adding" a "whole" web page to Hypothes.is, without annotating it ?

I tend to use page notes to do this so that they're not linked to a specific annotation but the whole page - I tag it with a bookmark tag (side note: would be cool to be able to search for page annotations to bulk tag with bookmark instead of doing it manually, this is something I'll have to remember to implement along with the local annotation switch). You could also check in the annotation template if a quote exists and if not, treat it as a page note and format it differently. 

Let me know how it goes - if you find any bugs of course, but also if there's any friction in the workflow that could be improved.

**Reply by ngirard** [reply]
> I tend to use page notes to do this so that they're not linked to a specific annotation but the whole page - I tag it with a bookmark tag

Aha, thanks, I totally forgot about page notes.

**Reply by ngirard** [reply]
> would be cool to be able to search for page annotations to bulk tag with bookmark instead of doing it manually

Could you please describe your manual process ?

> this is something I'll have to remember to implement along with the local annotation switch

How about filing an issue for this ?

**Reply by ngirard** [reply]
> You could also check in the annotation template if a quote exists and if not, treat it as a page note and format it differently.

Insightful, I'd have never thought of that !

I feel like creating new discussions about best practices, whose contents could later be used as the basis for more formal documentation or wiki pages. What do you think ?

**Reply by ngirard** [reply]
Incidentally, the GH discussions pages seem a bit buggy. For instance, that latest comment of yours didn't appear until I refresh the page.

**Reply by Ninjani** [reply]
> Could you please describe your manual process ?

Oh, I just add the bookmark tag to the page note in the browser sidebar.

> How about filing an issue for this ?

Yes, will do

> I feel like creating new discussions about best practices, whose contents could later be used as the basis for more formal documentation or wiki pages. What do you think ?

Great idea! It would also be inspiring I think to users who aren't super familiar with handlebars syntax yet (like me!).

> Incidentally, the GH discussions pages seem a bit buggy. For instance, that latest comment of yours didn't appear until I refresh the page.

Yes, same for me.

**Reply by ngirard** [reply]
So my work took over the rest, but I'll keep adding comments about my experiments, perhaps at a slow pace.

I've been toying with replacing page bookmarks with Hypothes.is' page annotations, and it looks promising.
I had one annoyance though: when in `gooseberry search`'s interactive mode, the search is performed against the page url but not its title. Perhaps it's the case with all the annotations and I didn't notice because it wasn't a problem ?

**Reply by ngirard** [reply]
Another thing I'm a bit embarrassed to report, because as a developer I can appreciate the efforts it would require.

I'm not keeping a blog or anything, but I regularly transmit pieces of information (collected during my daily tech monitoring) to the coworkers and relatives I think would find interesting.
`gooseberry make` would be the perfect match for this, except that I've geared the configuration to digesting my annotations in Obsidian, whereas transmitting them around would require a different configuration.

So I think it would be useful to be able to create different "publishing" configurations and tell `gooseberry search` which one to pick.
I'm afraid that would require a bit of work to implement though...

**Reply by Ninjani** [reply]
> the search is performed against the page url but not its title.

Could you perhaps make an issue for this? Should be an easy fix.

> So I think it would be useful to be able to create different "publishing" configurations and tell `gooseberry search` which one to pick.

Would the fix detailed in #54 work for this?

> I'll keep adding comments about my experiments, perhaps at a slow pace.

Okay, great! I'll try to merge this PR over the weekend, and then fix #85 and #92

**Reply by ngirard** [reply]
Another thing, which concerns both "publishing to others" and "digesting via Obsidian" workflow.

Once `gooseberry make <query>` is done, I wish I would tag the concerned annotations, so that I can distinguish the annotations that went through a workflow from the other ones, and only search the latter ones later.

For tagging these annotations with e.g. a `done` tag, I could use `gooseberry tag <query> -- done`, but it feels like `gooseberry make` could gain a dedicated options, such as `--then-tag done`. Not indispensable, though.

But then, for searching the annotations that aren't tagged `done`, I'm unsure whether it's currently possible using the `--tags` option. Reading the CLI help suggests it's not.

**Reply by ngirard** [reply]
> Could you perhaps make an issue for this? Should be an easy fix.

Sure ! Done in #93.

**Reply by ngirard** [reply]
> Would the fix detailed in #54 work for this?

Sure, that's Good Enough ® ! Thanks !

**Reply by Ninjani** [reply]
> But then, for searching the annotations that aren't tagged `done`, I'm unsure whether it's currently possible using the `--tags` option. Reading the CLI help suggests it's not.

No, not right now but I'll add this to #92. Will think a bit more about adding the tag option to `make` though, since it feels like the export won't actually reflect the annotations (since the exported markdown won't have the done tag but the annotations will)

> Sure ! Done in #93.

Thanks!

**Reply by ngirard** [reply]
> No, not right now but I'll add this to #92. 

Great, thanks !

> Will think a bit more about adding the tag option to `make` though, since it feels like the export won't actually reflect the annotations (since the exported markdown won't have the done tag but the annotations will)

Sure. Designing a good CLI is a process. Better take your time than doing it quick & dirty.

**Comment by Ninjani** [comment, marked as answer]
Done in v0.9.0!
